lib/tracing-limit/src/lib.rs

```diff
@@ -10,199 +9,204 @@ use std::{
+    }
 
-                let events = self.events.read().unwrap();
-                if let Some((count, start)) = events.get(&id) {
-                    let local_count = count.fetch_add(1, Ordering::Relaxed);
-                    let start = start.load(Ordering::Relaxed);
-
-                    if ts - start < limit {
-                        // If this log is being limited and we've seen it for
-                        // a second time within that window. We will log out that
-                        // we are now rate limiting this log.
-                        //
-                        // If we have seen this event more than twice,
-                        // then lets early return to avoid passing this
-                        // event to the inner subscriber.
-                        if local_count == 1 {
-                            let mut visitor = FmtVisitor::default();
-                            event.record(&mut visitor);
-
-                            let message = if let Some(message) = &visitor.message {
-                                &message
-                            } else {
-                                "unknown event"
-                            };
-
-                            let meta = event.metadata();
-
-                            println!(
-                                "{} {} Limiting the output of \"{}\" for the next {} seconds",
-                                FmtLevel(&tracing_core::Level::INFO),
-                                meta.target(),
-                                message,
-                                limit
-                            );
-                            return;
-                        } else if local_count > 1 {
-                            return;
-                        }
-                    } else {
-                        drop(events);
-
-                        let mut events = self.events.write().unwrap();
-                        events.remove(&id);
-                        drop(events);
-
-                        let meta = event.metadata();
-
-                        let mut visitor = FmtVisitor::default();
-                        event.record(&mut visitor);
-
-                        let message = if let Some(message) = &visitor.message {
-                            &message
+    fn enabled(&self, metadata: &Metadata, ctx: Context<S>) -> bool {
+        if is_limited(metadata) {
+            let id = metadata.callsite();
+            let events = self.events.read().expect("lock poisoned!");
+
+            // check if the event exists within the map, if it does
+            // that means we are currently rate limiting it.
+            if let Some(state) = events.get(&id) {
+                let ts = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_secs() as usize;
+
+                if ts - state.start < state.limit {
+                    if state.count.load(Ordering::SeqCst) == 1 {
```

Comment:
I don't think loads and stores to the counter need to [be part of the global memory order](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.SeqCst). You can relax this guarantee to `Acquire` for the loads and `Release` for the stores.
```suggestion
                    if state.count.load(Ordering::Acquire) == 1 {
```